### PR TITLE
feat(go): runtime object creation

### LINKED ISSUE #2039: Go - Define Runtime Object Creation Lifecycle
Now that go code generation is generating all types with correct type signatures, the generated constructor methods can call to the runtime library to instantiate objects within the JSII node runtime.

The first step in this process is defining runtime functions for object creation that are called within code generated "class" constructors. This will create a `CreateRequest` struct, serialize it, and pass it to the runtime client. It also will handle the response and deserializing any data within the returned JSON `CreateResponse` struct to the expected type of the generated function.

### LINKED ISSUE #2091: Go runtime investigate replacing JsiiObj Interface
The go runtime currently uses an interface that all jsii objects must implement in order to link instance IDs within the kernel to the host object instance itself. This isn't ideal since this interface can conflict user defined code. IE the methods of the interface may prevent users from defining functions with conflicting names.

Alternatively, we should investigate defining a global hash-map that maps reference IDs to object instances. This allows us to look up objects by reference ID and find reference IDs from instances, similarly to the current `getJsiiObjId` method used in runtime calls.

### PR DESCRIPTION
**Pull request by MrArnoldPalmer**
Defines go runtime library function for object creation. Adds `objects`
map to client struct in order to track object references after creation.
Changes object creation code gen implementation to call runtime
create function.

Adds `fqn` field to `GoType` to easily access fqn in generated runtime
function calls.

Fixes #2091
Fixes #2039

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/@jsii/go-runtime/jsii-experimental/runtime.go [hunk 1]
```diff
@@ -36,6 +36,26 @@ func Load(name string, version string, tarball []byte) {
 	}
 }
 
+// Create will construct a new JSII object within the kernel runtime. This is
+// called by jsii object constructors.
+func Create(fqn string, args []interface{}, interfaces []string, overrides []Override, returns interface{}) {
```
Comment (RomainMuller) [inline packages/@jsii/go-runtime/jsii-experimental/runtime.go:41]: I wonder if we should alias `string` to `FQN` in order to make the type name parameters more expressive?

File: packages/jsii-pacmak/lib/targets/go/runtime/index.ts [hunk 2]
```diff
@@ -60,22 +65,37 @@ export class ClassConstructor {
...
-    code.line(`Method: "Constructor",`);
-    code.line(
-      `Args: ${paramsString(
-        this.parent.parent.type.initializer!.parameters.map((p) =>
-          p.type.toString(),
-        ),
-      )},`,
-    );
-    code.close(`})`);
+    code.line(`returnval := ${this.parent.parent.name}{}`);
```
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/runtime/index.ts:68]: So we need to actually use a name here that will never conflict with any parameter of the constructor... Either some name that just cannot be a parameter name, or slugify appropriately if a parameter matches...

---

Alternatively (and perhaps this is a batter option?), make `JSII_CREATE_FUNC` return the `returnval`, so you can (⚠️ heavy schematization ahead):
```go
return JSII_CREATE_FUNC(
    "parent.parent.fqn",
    []args{},
    []interfaces{},
    []overrides{},
    &parent.parent.name{}
)
```

File: packages/jsii-pacmak/lib/targets/go/runtime/index.ts [hunk 3]
```diff
@@ -60,22 +65,37 @@ export class ClassConstructor {
...
+      .join(', ');
+    return `[]interface{}{${argsList}}`;
   }
 }
 
 export function emitInitialization(code: CodeMaker) {
   code.line(`${JSII_INIT_ALIAS}.${JSII_INIT_FUNC}()`);
 }
+
+export * from './constants';
```
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/runtime/index.ts:101]: Any chance we can move this up? Or better yet, export it from an index instead?

File: packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts [hunk 4]
```diff
@@ -97,6 +96,6 @@ export class GoTypeRef {
 
     // type isn't handled
     // TODO: Are there other cases to handle? if not throw an error.
```
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts:98]: Perhaps drop a warning so you know something unexpected is about to happen?

**Review by RomainMuller — CHANGES_REQUESTED** [review]

File: packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts (hunk 4 above)
Comment (MrArnoldPalmer) [reply packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts]: Added a debug message, using `log.warn` makes this annoying the the interim =).

**Review by RomainMuller — APPROVED** [review]

**Pull request merged**